New input implementation

mucaho opened the pull request:
Slim version #1116, as mentioned in comments there.

starwed commented on src/controls/controls-system.js:
```diff
@@ -89,6 +85,12 @@ Crafty.s("Controls", {
+        Crafty.s("Mouse").bind("MouseDown", this.updateTriggers.bind(this));
+        Crafty.s("Mouse").bind("MouseUp", this.updateTriggers.bind(this));
+        Crafty.s("Mouse").bind("DoubleClick", this.updateTriggers.bind(this));
+        Crafty.s("Mouse").bind("Click", this.updateTriggers.bind(this));
```
These are also bound in the `events` property to the global events.

Is the issue here that the Mouse system now does not trigger the events globally?  If so, should they be removed from the event block?

mucaho replied:
> Is the issue here that the Mouse system now does not trigger the events globally?

Exactly, if Mouse system would trigger the events globally, all mouse entities would receive it. But maybe that's not an issue anymore with the newly proposed implementation?

starwed commented on src/controls/inputs-keyboard.js:
```diff
@@ -0,0 +1,276 @@
+            changeFocus: function (oldFocusElem, newFocusElem) {
+                // TODO enable these once component iteration improved
+                //Crafty.s('Keyboard').trigger('FocusChange', newFocusElem);
+                //Crafty('Keyboard').trigger('FocusChange', newFocusElem);
```
This is somewhat related to an idea I originally had when working on systems -- giving a system a way of triggering events on a set of related components (by default with the same name).

In the end I decided that it was simpler to just always trigger events globally, and trust that any entity that subscribed to the event should receive it.  What's the specific rationale for only triggering a subset here (and I think elsewhere in the PR?)

starwed commented on src/controls/inputs-lifecycle.js:
```diff
@@ -0,0 +1,68 @@
+                        'DOMMouseScroll'; // old Firefox
+
+//initialize the input events onload
+Crafty._preBind("Load", function () {
```
Nice that these are broken out.

There might be a way to completely avoid using `_preBind` here, and have the systems handle the event binding and unbinding directly.  I don't recall the exact issue, but the comment for prebind says:
> There are some events that need to be bound to Crafty when it's started/restarted, so store them here
>  Switching Crafty's internals to use the new system idiom should allow removing this hack

starwed replied:
For reference, prebind was added in #960, which provides some more context as to why it was necessary.

starwed reviewed (commented):
A lot of the diffs here are code moving around, which makes it a little hard to read through.  Could you highlight what you consider the important changes here?

I tried to [put together a chart](https://www.lucidchart.com/documents/view/95ac4e27-ac53-4640-bb29-a6ed632f5427) that shows the current structure of keyboard inputs/controls, and shows my idea of where locking would occur.  (Let me know if you can't access.)  A big blocker for my idea is that we don't implement mapping pointer events to controls.  I don't want that to hold up this work, but I do want to see if we agree on a possible final goal here.  :)

mucaho commented:
Thank you for the feedback!   


>  A big blocker for my idea is that we don't implement mapping pointer events to controls _[yet]_

Yep, that's what I was wondering, too. If pointer events are being routed through the control system, the input system does not longer know which specific mouse entity should receive the event. Is this ok and should we just trigger mouse events globally? Anyways I'd leave this task to a subsequent PR.

mucaho commented:
> Could you highlight what you consider the important changes here?

The changes are mostly non-functional, in preparation for upcoming PRs that will shift the routing of events to the control system, introduce pointer events into the control system and add locking behavior to controls.
As you mentioned, most of the stuff is moving code around into appropriate namespaces. So instead of having `Crafty.resetMouseButtons`, `Crafty.mouseButtonsDown` and `Crafty.mouseDispatch` they would be properties / methods of the Mouse system. (Analogue for Keyboard, Touch).
The notion of dispatching and preventing the bubbling of events is abstracted into a common EventDispatcher base system that all these systems inherit from.
The validation of mousebutton- and keypresses is also abstracted into a common base system that these systems inherit from.
ResetMouseButton method is introduced, which similarly to resetKeyDown, resets actively pressed buttons once Crafty stage looses focus. These two methods are invoked on Crafty pause and stop, too.

I think your idea aligns with the goals of this PR. Will proceed with implementation and note detailed changes once done.

starwed commented:
>So instead of having Crafty.resetMouseButtons, Crafty.mouseButtonsDown and Crafty.mouseDispatch they would be properties / methods of the Mouse system

Makes sense.  Do we want to map the old names to the new system properties through some sort of alias, as done in #1071?

>If pointer events are being routed through the control system, the input system does not longer know which specific mouse entity should receive the event. Is this ok and should we just trigger mouse events globally? 

I think that would be the way to go, but to my mind there's a lot of uncertainty that can only really be pinned down by trying an implementation.

> Anyways I'd leave this task to a subsequent PR.

Agreed.

mucaho committed:
Add KeyboardState and MouseState templates

These templates abstract out the notion of validating incoming
key and mouse button (de)presses. Thus they offer
triggerKey / triggerMouse methods which propagate the incoming
event to the default trigger method, if the action is valid
(e.g. don't propagate KeyUp if KeyDown wasn't received beforehand).

That means that the methods Crafty.resetKeyDown and
Crafty.resetMouseDown can be called as often as needed using the
provided triggerKey / triggerMouse methods (e.g. due to
loss of window focus) without any unwanted side effects.

Additionally they provide convenience methods for querying the
pressed state of keys and buttons - isKeyDown / isButtonDown and
resetKeyDown / resetButtonDown.

MouseState tracks the last mouse event, too, which can be obtained
through the .lastMouseEvent property.

mucaho committed:
Streamline input implementation

* Split source and tests files into appropriate files:
src/controls/input is split into different subfiles
in the src/inputs directory. tests/unit/controls/input
is moved to tests/unit/inputs/input
* **API change:** Changes the signature and functionality of
Crafty.findPointerEventTargetByComponent:
This method now only finds the entity. A newly introduced
Crafty.augmentPointerEventCoordinates augments pointer events
with coordinates from real world space.
* Extract common functionality of dispatchers
(e.g. preventDefault) into a EventDispatcher prototype.
* **API change:** The various dispatch methods are converted
to systems. These systems inherit KeyboardState/MouseState,
as well as the EventDispatcher functionality. Additionally,
code is refactored to foster easier maintainance.
* All mouse events are triggered on the MouseSystem itself.
MouseEvents without target are no longer dispatched to a
specific system. MouseDrag, Viewport.mouseLook and
MouseTriggers from the ControlSystem bind to MouseSystem
directly in order to receive mouse events.
* **API aliasing:** Several properties and methods no longer
live under global Crafty namespace and are moved to appropriate
systems. Aliases are defined for those cases.
* Trigger Click event in touchDispatch if distance from start
to end of touch is smaller than some threshold.
* Augment MouseScroll event with real coordinates.
* Add resetMouseDown which similary to resetKeyDown triggers
ButtonUp events for currently pressed buttons.
ResetKeyDown and resetMouseDown are called on 3 occasions -
on Crafty.pause, on Crafty.stop and on stage focus loss.

mucaho commented:
Implementation is done! Commit messages should contain detailed messages.

------------------------------------

EDIT: removed outdated info from this comment

---------------------

> There might be a way to completely avoid using _preBind here, and have the systems handle the event binding and unbinding directly.

I would tackle that in another PR, if that's ok.

> Do we want to map the old names to the new system properties through some sort of alias

Done.

mucaho commented:
> I tried to put together a chart that shows the current structure of keyboard inputs/controls ... In the end I decided that it was simpler to just always trigger events globally, and trust that any entity that subscribed to the event should receive it. 

Done! Found a clean way of implementing it, which I had trouble with at first.

mucaho commented:
>> If pointer events are being routed through the control system, the input system does not longer know which specific mouse entity should receive the event. Is this ok and should we just trigger mouse events globally?
>
> I think that would be the way to go

I may have an idea to get best of both worlds:    
We let the input system find the clicked entity (if it exists) and save it as a e.g. `target` property on the mouse event. The mouse events still get dispatched globally, but user can check if the current `Mouse` entity was clicked by comparing it with that property. Otherwise, if we don't find the `target` for the user, each `Mouse` entity would have call `findPointerEventTargetByComponent` and compare itself to the result, which would be very inefficient.
However, the controls system still remains in charge of routing the thereafter produced Control event, either to the locked / focused element or in its absence to the suggested `target` from the mouse event.

mucaho commented:
(Just updated this PR for new "UpdateFrame" event, other than that no changes).
Interestingly, bug #1119 occurred on AppVeyor this time. Have a look at the status checks.

mucaho commented:
To summarize:
* This commit fixes part of #552, quote: "Someone could hold down a key
while we've lost focus, and then release once it's back, so we probably
don't want to fire keyup events in such a case."
* This commit fixes Click on touch devices (as reported in #1043).

I hope I've addressed all of your suggestions, merging this in, and follow-up PR coming soon.

mucaho commented:
I've updated the list of things to do in the [Controls System project](https://github.com/craftyjs/Crafty/projects/1). If you have some pointers, I'd appreciate the feedback!
